### ISSUE TITLE
Marshal ASN.1 OBJECT IDENTIFIER then unmarshal causes unmarshal error

### ISSUE THREAD

**Issue by sebastien-rosset** [golang/go#38737]
### What version of Go are you using (`go version`)?

go version go1.14.2 linux/amd64

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/serosset/.cache/go-build"
GOENV="/home/serosset/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/serosset/goroot"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build216516397=/tmp/go-build -gno-record-gcc-switches"
```
### What did you do?

Link to program:
https://play.golang.org/p/tI2WDW2A8Wm

### What did you expect to see?

1. Encode Object Identifier "1.2.36.67006527840.666.66.1.1" in DER format using the asn1.Marshal function.
1. Verify manually the byte encoding is correct
1. Take the ASN.1 encoded object identifier and try to unmarshal. Unmarshal should succeed and the same original object identifier should be retrieved.

### What did you see instead?

Step (4) is failing. asn1.Unmarshal fails with error **asn1: structure error: base 128 integer too large**

**Comment by odeke-em** [comment]
Thank you for this report @sebastien-rosset!

Kindly pinging the crypto squad @katiehockman @FiloSottile

**Comment by rolandshoemaker** [comment]
Looks like the original choice made by @agl was to match behavior on all platforms, so the maximum size was limited to 31 bits so that parsing behavior would match on 32 bit and 64 bit builds. In theory this could be supported by simply seeing if casting the parsed int64 -> int caused an overflow and only use the int if it didn't, which was the originally proposed solution to the problem, but this would cause mismatching behavior.

Confusingly this does create a marshal/unmarshal mismatch. asn1.Marshal will happily encode a integer larger than 31 bits, but cannot parse it. It seems like either parsing should be dependent on platform integer size, or marshal/unmarshal should match by limiting encoding to 31 bits on all platforms.

**Comment by sebastien-rosset** [comment]
> Looks like the original choice made by @agl was to match behavior on all platforms, so the maximum size was limited to 31 bits so that parsing behavior would match on 32 bit and 64 bit builds. In theory this could be supported by simply seeing if casting the parsed int64 -> int caused an overflow and only use the int if it didn't, which was the originally proposed solution to the problem, but this would cause mismatching behavior.

The golang ASN.1 implementation correctly implements the ASN.1 INTEGER type, which can have an arbitrary number of bits: An ASN.1 INTEGER can be written to an int, int32, int64, or *big.Int (from the math/big package).

An object identifier is specified in ITU-T recommendation X.208, which has been superseded bu  X.680-683. An object identifier value is semantically an ordered list of object identifier component values:
```
ObjIdComponent ::=
  NameForm |
  NumberForm |
  NameAndNumberForm

NumberForm ::= number | DefinedValue
```

Where `number` can have an arbitrary number of digits. It is not limited to 32 bits and it is not limited to 64 bits either. So it seems that to properly implement unmarshaling of object identifiers, the implementation should do something similar to INTEGER unmarshaling, where the value can be represented either as int32, 

> 
> Confusingly this does create a marshal/unmarshal mismatch. asn1.Marshal will happily encode a integer larger than 31 bits, but cannot parse it. It seems like either parsing should be dependent on platform integer size, or marshal/unmarshal should match by limiting encoding to 31 bits on all platforms.

**Comment by rolandshoemaker** [comment]
Sure, the problem though is that `asn1.ObjectIdentifier` was defined as a `[]int`, presumably because in the real world it's quite rare that components are ever >int32, and _extremely_ rare that components are >int64. That isn't to say that this doesn't happen, but for whatever reason that was the decision made.

As such this cannot be changed to something that supports arbitrary integer sizes without changing the definition, which would break _a lot_ of software that assumes the definition is `[]int`.

**Comment by sebastien-rosset** [comment]
Yes, I understand the challenges of the implementation. However I don't think just because `asn1.ObjectIdentifier` was defined as a `[]int`, that should mean golang will never be able to parse real-world CA certificates such as the New Zealand example I have provided.

**Comment by sebastien-rosset** [comment]
I don't think a solution would involve changing the type of ObjectIdentifier. As you and I pointed out, that would break a lot of code. But it seems this could be done in a way similar to what has been done for integers: the `asn1.UnmarshalWithParam()` function can take int, int32, int64 and big.Int for the `val` argument. Similarly, two new types `ObjectIdentifierInt64` and `ObjectIdentifierBigInt` could be defined and the  function `asn1.UnmarshalWithParam()` could have a couple more cases under the switch statement. The `asn1.Marshal()` function would also have to accept these two new types as input. Optionally there could be a wrapper type. That would cover pure ASN.1 operations.

Under the crypto package, I see 80 use of `asn1.ObjectIdentifier`.

**Comment by mikebros** [comment]
Hi, @sebastien-rosset. I represent the company that is having this issues. Thank you for posting this issues and looking for a resolution.
Can I ask you to please remove the references to our company name and function as we are not comfortable having us directly referenced in this bug report?
Thanks
